src/twisted/web/static.py

```diff
@@ -274,11 +273,33 @@ def directoryListing(self):
 
     def getChild(self, path, request):
         """
-        If this L{File}'s path refers to a directory, return a L{File}
+        If this L{File}"s path refers to a directory, return a L{File}
         referring to the file named C{path} in that directory.
 
-        If C{path} is the empty string, return a L{DirectoryLister} instead.
+        If C{path} is the empty string, return a L{DirectoryLister}
+        instead.
+
+        @param path: The current path segment.
+        @type path: L{bytes}
+
+        @param request: The incoming request.
+        @type request: An that provides L{twisted.web.iweb.IRequest}.
+
+        @return: A resource representing the requested file or
+            directory, or L{NoResource} if the path cannot be
+            accessed.
+        @rtype: An object that provides L{resource.IResource}.
         """
+        if isinstance(self.path, unicode):
+            try:
+                # Request calls urllib.unquote on each path segment,
+                # leaving us with raw bytes.
+                path = path.decode('utf-8')
```

Comment:
@markrwilliams This part of the patch is wrong, and breaks several builders.
`unicode` does not have a `.decode` method

Reply:
Whoops!  Thanks!

Reply:
These are the tests that appear to break:

https://github.com/twisted/twisted/blob/trunk/src/twisted/web/test/test_tap.py#L97

https://github.com/twisted/twisted/blob/trunk/src/twisted/web/test/test_tap.py#L107

https://github.com/twisted/twisted/blob/trunk/src/twisted/web/test/test_tap.py#L120

They pass native strings as the `path` argument to `Resource.getChild`, which is wrong.  URL paths in twisted.web are always bytes:

https://github.com/twisted/twisted/blob/trunk/src/twisted/web/http.py#L803-L818

https://github.com/twisted/twisted/blob/trunk/src/twisted/web/server.py#L168-L197

`FilePath`s can describe either unicode _or_ bytes paths. `static.File` is a subclass of `FilePath`, so `self.path` here can be either unicode or bytes. This is supposed to decode the request path only if `self.path` describes a unicode path; the original code is correct, `.decode` needs to be there, and those tests need to use bytes instead of native strings for their request paths.

Reply:
@markrwilliams You approved this PR.  Is it OK to merge this PR as is, and fix the test_tap.py usages in a follow-up PR, or do you want everything fixed in this PR?